salt/_pillar/metalk8s_nodes.py

```diff
@@ -37,6 +37,30 @@ def node_info(node, ca_minion):
             if role:
                 roles.add(role)
 
+    # Introduced as a workaround in cases which bootstrap setup fails after the
+    # kubernetes API server is up and running but node role labels are still
+    # not set.
+    #
+    # The issue is that when the API server is running the ext_pillar will
+    # overwrite the hardcoded roles with the ones from the API. It will cause
+    #  bootstrap setup failure as described here:
+    # https://github.com/scality/metalk8s/issues/2137
+    #
+    # The `is_bootstrap` flag is set in pillar only when run the
+    # `metalk8s.orchestrate.bootstrap` state.
+    if not roles and pillar.get("is_bootstrap", False):
+        if "metalk8s" in pillar:
+            if "nodes" in pillar["metalk8s"]:
+                if ca_minion in pillar["metalk8s"]["nodes"]:
+                    for role in pillar["metalk8s"]["nodes"][ca_minion].get("roles", []):
+                        log.info(
+                            "Workaround: Adding role '%s' to pillar in node '%s'",
+                            role,
+                            node["metadata"]["name"],
+                        )
+                        roles.add(role)
```

Comment:
To me you should update a bit the `node_info` function so that it take `is_bootstrap` and `pillar_roles` directly in argument, so that you send those value from the caller directly instead of the full pillar

Reply:
Here is the place we call the `node_info()` function.

```
        else:
            log.debug("Successfully retrieved nodes for ext_pillar")
            pillar_nodes = dict(
                (node["metadata"]["name"], node_info(node, ca_minion, pillar))
                for node in node_list
            )
```

In my opinion, I prefer to keep this part simple by passing the pilar.